Detection of 'native' CPU

fxcoudert opened the pull request:
Here's a first attempt calling `sysctlbyname("hw.cpufamily", …)`
At some point, we might want to fill in the blanks for older models, but definitely no hurry :)

iains committed:
libitm: support for aarch64-apple-darwin

Add arm64 as a target synonym and adjust the SjLj code to allow Mach-O
assembler.

libitm/ChangeLog:

	* config/aarch64/sjlj.S: Adjust for Mach-O.
	* configure.tgt: Allow arm64 as aarch64.

iains committed:
testsuite, aarch64, Darwin: add tests for aarch64-apple-darwin

	* gcc.target/aarch64/darwin/aarch64-darwin.exp: New darwin tests.
	* gcc.target/aarch64/darwin/complex-in-regs.c: New test.
	* gcc.target/aarch64/darwin/darwinpcs-d1.c: New test.
	* gcc.target/aarch64/darwin/darwinpcs-d2-00.c: New test.
	* gcc.target/aarch64/darwin/darwinpcs-d2-01.c: New test.
	* gcc.target/aarch64/darwin/darwinpcs-d2-02.c: New test.
	* gcc.target/aarch64/darwin/darwinpcs-d2-03.c: New test.
	* gcc.target/aarch64/darwin/darwinpcs-d2-04.c: New test.
	* gcc.target/aarch64/darwin/darwinpcs-d3.c: New (XFAIL) test.
	* gcc.target/aarch64/darwin/darwinpcs-d4.c: New test.
	* gcc.target/aarch64/darwin/homogeneous-aggr.c: New test.
	* gcc.target/aarch64/darwin/tu-accesses-0.c: New test.
	* gcc.target/aarch64/darwin/variadic-00.c: New test.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
aarch64, Darwin: Handle prfm/prfum with LLVM assembler.

The current LLVM-based assemblers reject offsets that are not suitable
for prfm as written in the local section. However, there is advice
elsewhere that says that this category of instruction should attempt
to use the 9bit unscaled version before falling back to the scaled one.

In the short-term reject values that the assembler will not accept.

This unpessimizes the prefetch cases for Darwin where the assembler
is not able to substitute the prfum instructions automagically.

gcc/ChangeLog:

	* config/aarch64/aarch64-protos.h (aarch64_address_valid_for_unscaled_prefetch_p):
	* config/aarch64/aarch64.cc (aarch64_address_valid_for_unscaled_prefetch_p):
	* config/aarch64/aarch64.md (prefetch_unscaled):
	* config/aarch64/constraints.md (Du):
	* config/aarch64/predicates.md (aarch64_unscaled_prefetch_operand):

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
c++, tls init: Temporary support for cross-tu static initialization [PR106435].

This patch set has been reviewed and there are suggestions for improvements
or a revised version.  However, in the short-term this is a usable placeholder.

===

1/ Currently, cross-tu static initialisation is not supported for targets without
alias support.

This patch adds support by building a stub function in place of the alias for
such targets.  Although this is slightly less efficient, it is a matter of
correcting wrong code and, for targets supporting sibcalls, actually the
penalty is a single jump for optimised code.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

	PR c++/106435

gcc/c-family/ChangeLog:

	* c-opts.cc (c_common_post_options): Allow fextern-tls-init for targets
	without alias support.

gcc/cp/ChangeLog:

	* decl2.cc (get_tls_init_fn): Allow targets without alias support.
	(handle_tls_init): Emit aliases for single init functions where the
	target supporst this, otherwise emit a stub function that calls the
	main tls init function.

gcc/testsuite/ChangeLog:

	* g++.dg/cpp0x/pr106435-b.cc: New test.
	* g++.dg/cpp0x/pr106435.C: New test.
	* g++.dg/cpp0x/pr106435.h: New test.

2/ C++, TLS, Darwin: Work around Darwin weak ref handling.

In order to handle cross-TU dynamic initialization of thread local variables,
we generate wrapper functions that call to the handler for the variable init.

Those inits are guarded so that they only run once.

From the perspective of a TU referencing a TLS variable, there is no way to
determine if it requires a guarded dynamic init.  So we make a weak reference
to said init and test for it at runtime.

This is fine at runtime for Darwin, but fails at link time when we have a TLS
var which does NOT need a guarded dynamic init (because then there is no init
and Darwin's ld64 does not permit undefined weak references at link time).

This patch is a work-around;  In the TU that is referencing the var, we now
generate both the wrapper _and_ a weak definition of a dummy init function.

So, in the case that there _is_ a dynamic init, the weak copy will be over-
ridden by the linker.  In the case that we have a trivial static init the
weak defined dummy init will be used.

TODO: rework this code-gen path to remove the test for the function (it will
always be true) since the combination of static and dynamic linkers will now
be responsible for correct operation.

iains committed:
libgcc, aarch64, Darwin: add DFP rules.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

libgcc/ChangeLog:

	* config.host: Add DFP rules.

iains committed:
configure: Allow libffi for aarch64-apple-darwin.

This is needed by rust.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

ChangeLog:

	* configure: Regenerate.
	* configure.ac: Permit libffi for aarch64 Darwin.

iains committed:
libgcc, aarch64: Add support for Darwin Aarch64.

Initial changes to build libgcc for aarch64 Darwin (declare the port,
handle Mach-O assembler syntax and arrange for building the main extra
pieces).

Darwin aarch64 does not exist for macOS 10.6, so there is no need to
build a shim to target it.

Enable heap trampolines.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

gcc/ChangeLog:

	* config/aarch64/aarch64.cc (TARGET_CUSTOM_FUNCTION_DESCRIPTORS):
	Choose the custom function trampoline bit number per subport.
	* config/aarch64/aarch64.h (AARCH64_CUSTOM_FUNCTION_TEST): Set the
	custom function bit number to 4.
	* config/aarch64/darwin.h: (AARCH64_CUSTOM_FUNCTION_TEST): Disable
	custom functions on Darwin (we use a trampoline on the heap).
	(CLEAR_INSN_CACHE): Call sys_icache_invalidate().

libgcc/ChangeLog:

	* config.host: Configure for aarch64*-*-darwin*.
	* config/aarch64/lse.S: Provide Mach-O version.
	* config/aarch64/t-darwin: New file.

fxcoudert committed:
aarch64, Darwin: Support IEEE754 Quad Precision as __float128.

aarch64-apple-darwin has a C long double type which is the same as the
double. In order to support a higher precision floating point type,
similarly to x86_64-apple-darwin, we define define __float128 (and
_Float128).

Co-authored-by: Iain Sandoe <iain@sandoe.co.uk>
Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

gcc/ChangeLog:

	* config/aarch64/aarch64-builtins.cc (AARCH64_BUILTIN_HUGE_VALQ,
	AARCH64_BUILTIN_INFQ): New built-ins.
	(aarch64_init_float128_types): New function...
	(aarch64_general_init_builtins): ... called here.
	(aarch64_general_fold_builtin): Initialize the new built-ins.
	* config/aarch64/aarch64.cc (aarch64_gimplify_va_arg_expr):
	Handle __float128 on darwin.
	(aarch64_mangle_type): Mangle __float128.
	(aarch64_c_mode_for_suffix): Assign suffix q to __float128.
	(aarch64_scalar_mode_supported_p): Ensure TFmode is allowed.
	* config/aarch64/aarch64.h (aarch64_float128_ptr_type_node): New
	node.
	* ginclude/stddef.h (__max_align_f128): Ensure it has correct
	alignment.

gcc/testsuite/ChangeLog:

	* g++.target/aarch64/float128-darwin-1.C: New test.
	* gcc.target/aarch64/darwin/float128-01.c: New test.

libgcc/ChangeLog:

	* config.host: Enable soft-float.
	* config/aarch64/sfp-machine.h: Provide definitions for
	aarch64-apple-darwin.
	* config/aarch64/t-darwin: Fix comment.

iains committed:
testsuite, aarch64, Darwin: Account for long-double is double.

Darwin's ABI has "long double" the same as "double" so that this test
does the same for the long double variants as the double ones and we need
to amend the counts accordingly.

gcc/testsuite/ChangeLog:

	* gcc.target/aarch64/xorsign.c: Amend scan-assembler times values
	for Darwin.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
testsuite, aarch64, darwin: Handle heap trampline case.

Darwin does not create the trampoline initializer in the TU, it is
done on demand by the builtins.

iains committed:
testsuite, aarch64, Darwin: Use __float128 instead of long double

Darwin's long double is the same as double, so use __float128 there to
test the same codegen as Linux.

iains committed:
testsuite, aarch64, Darwin: Skip a test that does not work for Darwin ABI.

Since Darwin indirects accesses to globals from other TUs, this test does
not work there.

iains committed:
testsuite, aarch64, Darwin: Allow unwind frames.

Darwin's function body scan code needs the function start and end
markers which are enabled by frames (either unwind or debug).  In the
absence of these, it's not clear how we'd delinate functions, since
Mach-O does not use .size.

gcc/testsuite/ChangeLog:

	* gcc.target/aarch64/tbz_1.c: Allow unwind frames for Darwin.
	* gcc.target/aarch64/tbz_2.c: Likewise.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
testsuite, aarch64, Darwin: Skip tests with incompatible stack layout.

Darwin's stack layout is sufficiently different from Linux that we really
need our own set of tests.  Skip these anyway.

gcc/testsuite/ChangeLog:

	* gcc.target/aarch64/stack-check-prologue-17.c: Skip for Darwin.
	* gcc.target/aarch64/stack-check-prologue-18.c: Likewise.
	* gcc.target/aarch64/stack-check-prologue-19.c: Likewise.
	* gcc.target/aarch64/stack-protector-8.c: Likewise.
	* gcc.target/aarch64/stack-protector-1.c: Skip for Darwin.
	* gcc.target/aarch64/stack-protector-2.c: Likewise.
	* gcc.target/aarch64/stack-protector-3.c: Likewise.
	* gcc.target/aarch64/stack-protector-4.c: Likewise.
	* gcc.target/aarch64/pr104689.c: Skip for Darwin.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
testsuite,aarch64,Darwin: Account for USER_LABEL_PREFIX and internal labels.

These tests were failing because they check for external functions and/or
local lables where the prefixes differ between Darwin and Linux.

gcc/testsuite/ChangeLog:

	* gcc.target/aarch64/pr103147-10.c: Account for Mach-o labels.
	* gcc.target/aarch64/pr109072_1.c: Likewise.
	* gcc.target/aarch64/reg-alloc-4.c: Allow Mach-O label syntax.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
testsuite,aarch64,Darwin: We have a signed char, so be specific about unsigned.

The test scans for an instruction that assumes unsigned chars (the default on Linux)
Darwin's is signed, so make the test specific about using unsigned.

gcc/testsuite/ChangeLog:

	* gcc.target/aarch64/mops_4.c: Make unsigned char explicit.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
testsuite,aarch64,Darwin: XFAIL because if large code model.

mcmodel=large is on the TODO, until then we expect this to fail.

gcc/testsuite/ChangeLog:

	* gcc.target/aarch64/pr63304_1.c: XFAIL for Darwin.
	* gcc.target/aarch64/pr78733.c: XFAIL for Darwin.
	* gcc.target/aarch64/pr79041-2.c: Likewise.
	* gcc.target/aarch64/pr63304_1.c: XFAIL Darwin.
	* gcc.target/aarch64/pr70120-2.c: Likewise.
	* gcc.target/aarch64/pr78733.c: Likewise.
	* gcc.target/aarch64/pr79041-2.c: Likewise.
	* gcc.target/aarch64/pr94530.c: Likewise.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
testsuite,aarch64,Darwin: Skip unimplemented patchable function tests.

It's not clear what we could do here, unless clang supports this too.

gcc/testsuite/ChangeLog:

	* gcc.target/aarch64/pr92424-2.c: Skip for Darwin.
	* gcc.target/aarch64/pr92424-3.c: Likewise.
	* gcc.target/aarch64/pr98776.c: Likewise.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
testsuite,aarch64,Darwin: Account for Mach-O relocation syntax.

Another case of :lo12:.LCxx c.f. lCxx@PAGEOFF.

gcc/testsuite/ChangeLog:

	* gcc.target/aarch64/vec-init-single-const.c: Allow Mach-O relocation
	syntax.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
testsuite,aarch64,Darwin: XFAIL tests that track per function attrs.

We have not implemented per-function tracking in Darwin so XFAIL these for
now.

gcc/testsuite/ChangeLog:

	* gcc.target/aarch64/target_attr_1.c: XFAIL for Darwin.
	* gcc.target/aarch64/target_attr_15.c: Likewise.
	* gcc.target/aarch64/target_attr_4.c: Likewise.
	* gcc.target/aarch64/target_attr_7.c: Likewise.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
aarch64, Darwin: Initial implementation of Apple CPU cores.

Fixes Issue #86.

This:
 * does not account for cores used in iOS etc.
 * proves to be quite tricky, since here is no published info (as far as
I know).

Sources of information;
 * XNU (only for the Apple Implementer ID)
 * sysctl -a | grep hw on various M1 and M2 machines
 * AArch64.td from the Apple Open Source repo for LLVM.
 * What XCode-14 clang passes to cc1.

Unfortunately, these sources are in conflict; in particular the clang-claimed
feature set disagrees with the output of sysctl -a, and the base Arm revs.
claimed in some cases miss features that ARM DDI 0487J.a lists as mandatory
for the rev.

GCC does not seem to have a scheduler that is similar to the "Cyclone" one
in LLVM - so I've guessed to use cortex57 (but, maybe we miss 8-issue, it's
not clear - and my experience with the scheduler is ≈ 0).

Thus, the choices here are intended to be conservative.

TODO:
 * Currently, we do not seem to have any way to specify that M2 has support
  for FEAT_BTI, but AFAICT is not compliant with the Arm base rev that implies
  this.
 * No real idea what the version numbers should be, and if there's anywhere
   we could actually read them.
 * Since we have FIRESTORM/ICESTORM and similar pairs for the performance and
   efficiency cores on various machines, perhaps we should be using a big.LITTLE
   configuration; OTOH currently, I have no idea if that is usable in any way
   with our port.

gcc/ChangeLog:

	* config/aarch64/aarch64-cores.def (AARCH64_CORE): Add Apple-a12,
	Apple-M1, Apple-M2.
	* config/aarch64/aarch64-tune.md: Likewise.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
testsuite,Darwin,aarch64: Omit FP in stack-check tests.

These specifically test for absent FP (but aarch64 Darwin mandates one).
Perhaps, we might better skip the tests.

	* gcc.dg/stack-check-5.c: Omit FP for aarch64 Darwin.
	* gcc.dg/stack-check-6.c: Likewise.
	* gcc.dg/stack-check-9.c: Likewise.
	* gcc.dg/stack-check-10.c: Likewise.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
D: disable collect fork on Darwin

iains committed:
D: disable collect fork on Darwin

iains committed:
testsuite, aarch64: Handle optional +crc.

We have two conflicting assembler bugs, one that requires +crc to be appended
and one that does not so we have to make it optional.

Some of these tests were intending to be testing that certain extensions are
not emitted, added scan-assembler-not for those.

gcc/testsuite/ChangeLog:

	* gcc.target/aarch64/options_set_1.c:i XFAIL for Darwin.
	* gcc.target/aarch64/options_set_11.c: Make +crc optional.
	* gcc.target/aarch64/options_set_12.c: Likewise.
	* gcc.target/aarch64/options_set_13.c: Likewise.
	* gcc.target/aarch64/options_set_14.c: Likewise.
	* gcc.target/aarch64/options_set_15.c: Likewise.
	* gcc.target/aarch64/options_set_16.c: Likewise.
	* gcc.target/aarch64/options_set_17.c: Likewise.
	* gcc.target/aarch64/options_set_18.c: Likewise.
	* gcc.target/aarch64/options_set_19.c: Likewise.
	* gcc.target/aarch64/options_set_2.c: Likewise.
	* gcc.target/aarch64/options_set_20.c: Likewise.
	* gcc.target/aarch64/options_set_21.c: Likewise.
	* gcc.target/aarch64/options_set_22.c: Likewise.
	* gcc.target/aarch64/options_set_23.c: Likewise.
	* gcc.target/aarch64/options_set_24.c: Likewise.
	* gcc.target/aarch64/options_set_25.c: Likewise.
	* gcc.target/aarch64/options_set_26.c: Likewise.
	* gcc.target/aarch64/options_set_3.c: Likewise.
	* gcc.target/aarch64/options_set_4.c: Likewise.
	* gcc.target/aarch64/options_set_5.c: Likewise.
	* gcc.target/aarch64/options_set_6.c: Likewise.
	* gcc.target/aarch64/options_set_7.c: Likewise.
	* gcc.target/aarch64/options_set_8.c: Likewise.
	* gcc.target/aarch64/options_set_9.c: Likewise.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
testsuite, aarch64, Darwin: XFAIL bfloat16-complex.c.

This fails as per PR 111876.

gcc/testsuite/ChangeLog:

	* gcc.dg/torture/bfloat16-complex.c: XFAIL for Darwin.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
aarch64: Make assembler bug workaround configurable.

Some assmblers have a bug that requires +crc to be emitted even
though the base architecture supports it.  However, that also
triggers a different bug in another assembler.  So make the fix
configurable.

gcc/ChangeLog:

	* common/config/aarch64/aarch64-common.cc: Make the asm
	crc bug workaround configurable.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
aarch64: Fix for PR111867.

From Andrew Pinski.

gcc/ChangeLog:

	* config/aarch64/aarch64.cc (aarch64_float_const_representable_p):

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
testsuite,aarch64,Darwin: Update tests after recent codegen changes.

The unfortunate property of function body matching is that it can be
sensitive to changes in codegen that are unrelated to the test iself.

This means checking that the test still works for the intended purpose
and reorganising the match clauses.

gcc/testsuite/ChangeLog:

	* gcc.target/aarch64/darwin/variadic-01.c: Compensate for recent
	codegen changes.
	* gcc.target/aarch64/darwin/variadic-02.c: Likewise.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
testsuite, aarch64: Specify tune=geneic for some tests.

The function body scans are dependent in some cases on the cost
models used for the code gen.  When the compiler is configured with
a non-default CPU this leads to failures.  So specify generic tune
and also omit frame pointers for Darwin.

gcc/testsuite/ChangeLog:

	* gcc.target/aarch64/ldp_stp_14.c: Specify generic tune and
	omit the frame pointer for Darwin.
	* gcc.target/aarch64/ldp_stp_15.c: Likewise.
	* gcc.target/aarch64/ldp_stp_16.c: Likewise.
	* gcc.target/aarch64/ldp_stp_18.c: Likewise.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
testsuite, Darwin: Add support for Aarch64 Darwin.

This is mostly mechanical changes to recognise the port triple and, in some
instances different assembler directives.

gcc/testsuite/ChangeLog:

	* g++.dg/torture/darwin-cfstring-3.C: Add aarch64-darwin support.
	* gcc.dg/builtin-apply2.c: Likewise.
	* gcc.dg/cwsc1.c: Likewise.
	* gcc.dg/pr26427.c: Likewise.
	* gcc.dg/pubtypes-2.c: Likewise.
	* gcc.dg/pubtypes-3.c: Likewise.
	* gcc.dg/pubtypes-4.c: Likewise.
	* gcc.dg/rtl/aarch64/big-endian-cse-1.c: Likewise.
	* gcc.dg/tls/pr78796.c: Likewise.
	* gcc.dg/torture/darwin-cfstring-3.c: Likewise.
	* gcc.dg/torture/stackalign/builtin-apply-2.c: Likewise.
	* gcc.dg/tree-ssa/stdarg-2.c: Likewise.
	* gcc.dg/tree-ssa/stdarg-4.c: Likewise.
	* gcc.dg/tree-ssa/stdarg-5.c: Likewise.
	* gfortran.dg/pr95690.f90: Likewise.
	* lib/target-supports.exp: Likewise.
	* obj-c++.dg/gnu-api-2-class-meta.mm: Likewise.
	* obj-c++.dg/gnu-api-2-class.mm: Likewise.
	* obj-c++.dg/torture/strings/const-cfstring-4.mm: Likewise.
	* obj-c++.dg/torture/strings/const-str-10.mm: Likewise.
	* obj-c++.dg/torture/strings/const-str-11.mm: Likewise.
	* obj-c++.dg/torture/strings/const-str-9.mm: Likewise.
	* objc.dg/gnu-api-2-class-meta.m: Likewise.
	* objc.dg/gnu-api-2-class.m: Likewise.
	* objc.dg/torture/strings/const-cfstring-4.m: Likewise.
	* objc.dg/torture/strings/const-str-10.m: Likewise.
	* objc.dg/torture/strings/const-str-11.m: Likewise.
	* objc.dg/torture/strings/const-str-9.m: Likewise.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
testsuite, aarch64, Darwin: Skip sve, sve2 and acle tests.

Darwin does not have support for these features; skipping the tests
seems the least testsuite maintainance.  If we build a cross on Darwin
then the tests will run as normal - testing that they run for Darwin as
a target seems to be a waste of cpu cycles.

If Darwin gets any of these features in a future chip - we can revisit.

gcc/testsuite/ChangeLog:

	* g++.target/aarch64/sve/aarch64-sve.exp: Skip for Darwin.
	* g++.target/aarch64/sve/acle/aarch64-sve-acle-asm.exp: Likewise.
	* g++.target/aarch64/sve/acle/aarch64-sve-acle.exp: Likewise.
	* g++.target/aarch64/sve2/acle/aarch64-sve2-acle-asm.exp: Likewise.
	* gcc.target/aarch64/aapcs64/aapcs64.exp: Likewise.
	* gcc.target/aarch64/sve/aarch64-sve.exp: Likewise.
	* gcc.target/aarch64/sve/acle/aarch64-sve-acle-asm.exp: Likewise.
	* gcc.target/aarch64/sve/acle/aarch64-sve-acle.exp: Likewise.
	* gcc.target/aarch64/sve/pcs/aarch64-sve-pcs.exp: Likewise.
	* gcc.target/aarch64/sve2/aarch64-sve2.exp: Likewise.
	* gcc.target/aarch64/sve2/acle/aarch64-sve2-acle-asm.exp: Likewise.
	* gcc.target/aarch64/sve2/acle/aarch64-sve2-acle.exp: Likewise.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
testsuite, aarch64, Darwin: Handle Mach-O assembly output.

Amend the tests to expect Mach-O syntax for the scan asm.

gcc/testsuite/ChangeLog:

	* g++.dg/abi/aarch64_guard1.C: Handle Mach-O.
	* g++.dg/abi/arm_va_list.C: Likewise.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
testsuite, aarch64, Darwin: Unsorted changes

iains committed:
call-lowering: Allow promote_function_mode to depend on cumulative args.

TODO: make a much longer explanation of the changes....

For at least one ABI the promotion rules for small integers (QI, HI)
depend on whether the argument will be passed in register or on the stack.

This change permits the target to make such an adjustment, with a default
implementation to behave the same as currently.

gcc/ChangeLog:

	* calls.cc (initialize_argument_information): Use argument
	promotion with cumulative args available.
	(emit_library_call_value_1): Likewise.
	* config/aarch64/aarch64.cc
	(aarch64_function_arg_alignment): Mach-O may have promoted argument
	modes when considering alignment.
	(aarch64_darwin_promote_fn_mode): New.
	(aarch64_darwin_promote_function_mode_ca): New.
	* config/aarch64/darwin.h (TARGET_PROMOTE_FUNCTION_MODE): New.
	(TARGET_PROMOTE_FUNCTION_MODE_CA): New.
	* doc/tm.texi: Document TARGET_PROMOTE_FUNCTION_MODE_CA.
	* doc/tm.texi.in: Likewise.
	* explow.cc (promote_function_mode): Add an overload taking a
	cumulative args input.
	* explow.h (promote_function_mode): New overload.
	* function.cc (assign_parm_find_data_types): Use argument
	promotion with cumulative args available.
	* target.def (PROMOTE_FUNCTION_MODE_CA): New hook.
	* targhooks.cc (default_promote_function_mode_ca): New.
	* targhooks.h (default_promote_function_mode_ca): New.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

t-promote-args-fixes

iains committed:
aarch64: Remove unusued prototype.

This has no corresponding function.

gcc/ChangeLog:

	* config/aarch64/aarch64-protos.h
	(aarch64_init_cumulative_incoming_args): Remove.

iains committed:
testsuite, aarch64, Darwin: Amend tests after D.3 fix.

Since Darwin promotes small integers passed in registers in the caller
rather than callee, some of the function body tests need updating to
handle the difference.

gcc/testsuite/ChangeLog:

	* gcc.target/aarch64/ands_3.c: Amend for darwinpcs.
	* gcc.target/aarch64/extend.c: Likewise.
	* gcc.target/aarch64/ldar_1.c: Likewise.
	* gcc.target/aarch64/movhi_1.c: Likewise.
	* gcc.target/aarch64/movqi_1.c: Likewise.
	* gcc.target/aarch64/tbz_1.c: Likewise.
	* gcc.target/aarch64/tbz_2.c: Likewise.
	* gcc.target/aarch64/tbz_3.c: Likewise.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
Darwin: Changes to add AArch64/Arm64 sub-port.

This is ....

gcc/ChangeLog:

	* config.gcc:
	* config/darwin-driver.cc (darwin_driver_init):
	* config/darwin.cc (machopic_legitimize_pic_address):
	(machopic_finish):
	(darwin_make_eh_symbol_indirect):
	(darwin_file_end):
	* config/darwin.h (DARWIN_ARM64):

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
aarch64: New files for the Darwin AArch64 sub-port.

gcc/ChangeLog:

	* config/aarch64/darwin.h: New file.
	* config/aarch64/darwinpcs.md: New file.
	* config/aarch64/t-aarch64-darwin: New file.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
aarch64: Changes to support the Darwin sub-port.

gcc/ChangeLog:

	* config/aarch64/aarch64-builtins.cc (enum aarch64_builtins):
	(aarch64_init_subtarget_builtins):
	* config/aarch64/aarch64-c.cc (aarch64_register_pragmas):
	* config/aarch64/aarch64-protos.h (enum aarch64_symbol_type):
	(aarch64_address_valid_for_unscaled_prefetch_p):
	(aarch64_init_subtarget_builtins):
	* config/aarch64/aarch64.cc (aarch64_elf_asm_destructor):
	(aarch64_hard_regno_mode_ok):
	(aarch64_load_symref_appropriately):
	(aarch64_expand_mov_immediate):
	(aarch64_return_in_memory):
	(aarch64_vfp_is_call_candidate):
	(aarch64_layout_arg):
	(aarch64_init_cumulative_args):
	(aarch64_function_arg_advance):
	(aarch64_function_arg_regno_p):
	(aarch64_function_arg_round_boundary):
	(aarch64_classify_address):
	(aarch64_address_valid_for_unscaled_prefetch_p):
	(output_macho_postfix_expr):
	(aarch64_print_operand):
	(aarch64_print_address_internal):
	(aarch64_asm_output_labelref):
	(aarch64_fold_builtin):
	(initialize_aarch64_code_model):
	(aarch64_classify_symbol):
	(GTY):
	(aarch64_build_builtin_va_list):
	(aarch64_expand_builtin_va_start):
	(aarch64_gimplify_va_arg_expr):
	(aarch64_setup_incoming_varargs):
	(aarch64_mangle_type):
	(aarch64_mov_operand_p):
	(aarch64_declare_function_name):
	(aarch64_asm_output_alias):
	(aarch64_start_file):
	(aarch64_c_mode_for_suffix):
	(aarch64_output_simd_mov_immediate):
	(aarch64_sls_emit_shared_blr_thunks):
	(TARGET_ASM_UNALIGNED_HI_OP):
	(TARGET_ASM_UNALIGNED_SI_OP):
	(TARGET_ASM_UNALIGNED_DI_OP):
	(TARGET_FUNCTION_ARG_ROUND_BOUNDARY):
	* config/aarch64/aarch64.h (TARGET_MACHO):
	(SUBTARGET_EXTRA_SPECS):
	(EXTRA_SPECS):
	* config/aarch64/aarch64.md (prefetch_unscaled):
	* config/aarch64/aarch64.opt:
	* config/aarch64/constraints.md (Du):
	* config/aarch64/predicates.md (aarch64_unscaled_prefetch_operand):

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
call lowering, aarch64, Darwin: Provide control over stack layout and packing.

1. Generic : allow different passing organisation for variadic funcs.

Right now, the generic code assumes that it can determine the stack layout from
target callbacks that don't have complete information about the argument they
are processing (they only know the mode and the type - not it the argument is
named or not, for example).

For aarch64-apple-darwin, we really need the option for a target to
override the layout process - if it requires distinct stack layouts for
different calling conventions.

2. Add cumulative_args_t variants of TARGET_FUNCTION_ROUND_BOUNDARY and friends

The two target hooks responsible for informing GCC about stack
parameter alignment are `TARGET_FUNCTION_ARG_BOUNDARY` and
`TARGET_FUNCTION_ARG_ROUND_BOUNDARY`, which currently only consider
the tree and machine_mode of a specific given argument.

Create two new target hooks suffixed with '_CA', and pass in a third
`cumulative_args_t` parameter. This enables the backend to make
alignment decisions based on the context of the whole function rather
than individual parameters.

The orignal machine_mode/tree type macros are not removed - they are
called by the default implementations of `TARGET_...BOUNDARY_CA` and
`TARGET_...ROUND_BOUNDARY_CA`. This is done with the intetnion of
avoiding large mechanical modifications of nearly every backend in
GCC. There is also a new flag, -fstack-use-cumulative-args, which
provides a way to completely bypass the new `..._CA` macros. This
feature is intended for debugging GCC itself.

3. Implement TARGET_..._CA target hooks for AArch64 Darwin

The AArch64 Darwin platform requires that named stack arguments are
passed naturally-aligned, while variadic stack arguments are passed on
word boundaries. Use the TARGET_FUNCTION_ARG_BOUNDARY_CA and
TARGET_FUNCTION_ARG_ROUND_BOUNDARY_CA target hooks to let the backend
correctly layout stack parameters.

4. aarch64, Darwin: Ensure we use Mach-O section selections.

At present, I have left in the ability to specify
pc-relative-literal-loads but it is not clear if there is a viable set
of relocatinos available to make use of this (or, indeed, if it is of
any use period).  However this is behind a command line option so we can
check up on it in slower time.

5. aarch64, Darwin: darwinpcs mandates a frame pointer.

However, clang _does_ omit the FP for leaf functions so we are going to
do the same.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

gcc/ChangeLog:

	* calls.cc (initialize_argument_information): Pass `args_so_far`.
	* calls.h (function_arg_info): Add `last_named` member.
	* common.opt: New flag `-fstack-use-cumulative-args`.
	* config.gcc: Set -fstack-use-cumulative-args to be off by default.
	Enable by default if the host platform is MacOS 11.x or 12.x and
	we're on AArch64.
	* target.h (cumulative_args_t): Move declaration from here, to...
	* cumulative-args.h (cumulative_args_t): ...this new file. This is
	to permit backends to include the declaration of cumulative_args_t
	without dragging in circular dependencies.
	* function.cc (assign_parm_find_entry_rtl): Provide
	cumulative_args_t to locate_and_pad_parm.
	(gimplify_parameters): Ditto.
	(locate_and_pad_parm): Conditionally call new hooks if we're
	invoked with -fstack-use-cumulative-args.
	* function.h: Include cumulative-args.h.
	(locate_and_pad_parm): Add cumulative_args_t parameter.
	* target.def (function_arg_boundary_ca): Add.
	(function_arg_round_boundary_ca): Ditto.
	* targhooks.cc (default_function_arg_boundary_ca): Implement.
	(default_function_arg_round_boundary_ca): Ditto.
	* targhooks.h (default_function_arg_boundary_ca): Declare.
	(default_function_arg_round_boundary_ca): Ditto.
	* doc/invoke.texi (-fstack-use-cumulative-args): Document.
	* doc/tm.texi: Regenerate.
	* doc/tm.texi.in: Ditto.
	* config/aarch64/aarch64-protos.h
	(aarch64_init_cumulative_incoming_args): Declare.
	* config/aarch64/aarch64.cc (aarch64_init_cumulative_args): Initialize
	`darwinpcs_n_named` (Total number of named parameters) and
	`darwinpcs_n_args_processed` (Total number of parameters we
	have processed, including variadic if any.)
	(aarch64_init_cumulative_incoming_args): Implement the
	INIT_CUMULATIVE_INCOMING_ARGS macro in order to capture
	information on the number of named parameters for the current
	function.
	(aarch64_function_arg_advance): Increment
	`darwinpcs_n_args_processed` each time we layout a function
	parameter.
	(aarch64_function_arg_boundary_ca): Implement
	TARGET_FUNCTION_ARG_BOUNDARY_CA and
	TARGET_FUNCTION_ARG_ROUND_BOUNDARY_CA to layout args based on
	whether we're a named parameter or not.
	(aarch64_function_arg_round_boundary_ca): Ditto.
	(TARGET_FUNCTION_ARG_BOUNDARY_CA): Define.
	(TARGET_FUNCTION_ARG_ROUND_BOUNDARY_CA): Ditto.
	(aarch64_darwin_frame_pointer_required): New.
	(TARGET_FRAME_POINTER_REQUIRED): New.
	(aarch64_select_rtx_section): Pick up Mach-O
	sections for Darwin.
	* config/aarch64/aarch64.h (CUMULATIVE_ARGS): Add
	`darwinpcs_n_named` and `darwinpcs_n_args_processed`.
	(INIT_CUMULATIVE_INCOMING_ARGS): Define.

+ aarch64, Darwin: Allow the user to use fomit-frame-pointer.

Although this will create objects that do not obey the strict ABI, we must
assume the user knows what they are doing.

This also refines somewhat the conditions for forcing a frame pointer.

gcc/ChangeLog:

	* config/aarch64/aarch64.cc
	(aarch64_darwin_frame_pointer_required): Allow fomit-frame-pointer.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
testsuite, Darwin: Amend dfp tests.

This adjusts for the USER_LABEL_PRIFIX and the counts that differ
for the Darwin implementation.

gcc/testsuite/ChangeLog:

	* gcc.dg/torture/convert-dfp-2.c: Handle Darwin.
	* gcc.dg/torture/convert-dfp.c: Likewise.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

fxcoudert committed:
First attempt

fxcoudert commented:
Some added notes on https://github.com/iains/gcc-darwin-arm64/commit/43a9cd85686ff5f7446a1129bc1ebf54826e9f59 (aarch64, Darwin: Initial implementation of Apple CPU cores).

- For Apple cores, we don't care about the `IMP` and `PART` arguments of the `AARCH64_CORE` macro. They're only used for core identification from /proc/cpuinfo, which we are bypassing.
- Alternatively, we could reuse these fields to carry the values of `hw.cpufamily` (M1, M2, etc) and `hw.cpusubfamily` (M1, M1 Pro, etc). But I think a simple switch is easier to read and maintain at this stage.

iains commented:
I was looking at the big.LITTLE stuff (I.e. where we have performance and efficiency cores), but I do not think that there's any way that we can action it at present.

As far as we might make action at present, the difference between M1 and M2 could be useful (M2 has support for BTI and BF16) but the BTI case is a bit complex since the aarch64 backend assumes it comes with armv8.5 and M2 is _mostly_ armv8.5 but it is missing one mandatory feature (according to the Arm ref. manual) so I've made it as armv8.4+additions.

iains committed:
aarch64, Darwin: Change the PAGEOFF format modifier from K to J.

Recent upstream backend changes have used K; Shift Darwin to use
J for this.

gcc/ChangeLog:

	* config/aarch64/aarch64.cc (aarch64_print_operand): Use J
	as the print operand modifier for PAGEOFF.
	* config/aarch64/aarch64.md: Likewise.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
libgcc, aarch64i, Darwin: Initial support for SME additions.

gcc/ChangeLog:

	* config/aarch64/aarch64.cc (aarch64_frame_pointer_required):
	* config/aarch64/darwin.h:

libgcc/ChangeLog:

	* config/aarch64/__arm_sme_state.S:
	* config/aarch64/libgcc-sme.ver:
	* config/aarch64/t-aarch64:

gcc/testsuite/ChangeLog:

	* gcc.target/aarch64/sme/aarch64-sme.exp:

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
aarch64, testsuite, darwin: update for improved ld/stp codegen

iains committed:
testsuite, Darwin: Deal with missing strndup in libc.

Earlier versions of macOS do not have the strn* functions in
libc, so we need to make parts of these tests conditional.

gcc/testsuite/ChangeLog:

	* gcc.dg/builtin-dynamic-object-size-0.c: Omit cases
	needing stndup, if the target does not support it.
	* gcc.dg/builtin-object-size-1.c: Likewise.
	* gcc.dg/builtin-object-size-2.c: Likewise.
	* gcc.dg/builtin-object-size-3.c: Likewise.
	* gcc.dg/builtin-object-size-4.c: Likewise.
	* gcc.dg/builtin-object-size-common.h: Set a flag
	to denote the availability of strndup.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
jit, Darwin: Implement library exports list.

Currently, we have no exports list for libgccjit, which means that
all symbols are exported, including those from libstdc++ which is
linked statically into the lib.  This causes failures when the
shared libstdc++ is used but some c++ symbols are satisfied from
libgccjit.

This implements an export file for Darwin (which is currently
manually created by cross-checking libgccjit.map).  Ideally we'd
script this, at some point.  Update libtool current and age to
reflect the current ABI version (we are not bumping the SO name
at this stage).

This fixes a number of new failures in jit testing.

gcc/jit/ChangeLog:

	* Make-lang.in: Implement exports list, and use a shared
	libgcc.
	* libgccjit.exp: New file.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
jit, Darwin: Implement library exports list.

Currently, we have no exports list for libgccjit, which means that
all symbols are exported, including those from libstdc++ which is
linked statically into the lib.  This causes failures when the
shared libstdc++ is used but some c++ symbols are satisfied from
libgccjit.

This implements an export file for Darwin (which is currently
manually created by cross-checking libgccjit.map).  Ideally we'd
script this, at some point.  Update libtool current and age to
reflect the current ABI version (we are not bumping the SO name
at this stage).

This fixes a number of new failures in jit testing.

gcc/jit/ChangeLog:

	* Make-lang.in: Implement exports list, and use a shared
	libgcc.
	* libgccjit.exp: New file.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
jit: Ensure ssize_t is defined.

One some targets it seems that ssize_t is not defined by any of the
headers transitively included by <stdio.h>.  This leads to a bootstrap
fail when jit is enabled.

The fix proposed here is to include sys/types.h when it is available
since that is where Posiz specifies that ssize_t is defined.

gcc/jit/ChangeLog:

	* libgccjit.h: Conditionally include <sys/types.h> on Darwin
	to ensure declaration of ssize_t.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
aarch64, Darwin: add libheapt_w to crts.

jicama committed:
build: drop target libs from LD_LIBRARY_PATH [PR105688]

The patch for PR22340 (r104978) moved the adding of TARGET_LIB_PATH to
RPATH_ENVVAR from POSTSTAGE1_HOST_EXPORTS to HOST_EXPORTS, but didn't
mention that in the ChangeLog; it also wasn't part of the patch that was
sent to gcc-patches.  I suspect it was included accidentally?

It also causes PR105688 when rebuilding stage1: once the stage1 libstdc++
has been built, if calling the system gcc to build host code involves
invoking any tool that links against libstdc++.so (gold, ccache) they get
the just-built library instead of the system library they expect.

Reverting that hunk of the change fixed my problem with bubblestrapping GCC
12 with ccache on a host with a newer system libstdc++.

But I believe that adding TARGET_LIB_PATH to RPATH_ENVVAR is not needed for
post-stage1 either.  Including TARGET_LIB_PATH goes back to r37545, with the
stated rationale of getting other C++ library configury to succeed, but that
seems like a rationale for having it in TARGET_EXPORTS, not HOST_.

And when building stage2+, looking for current stage target libraries isn't
what we want anyway; we would have wanted the previous stage target
libraries, and we haven't been looking for them, except for the previous
stage libgcc that we get from HOST_LIB_PATH_gcc.  We do want them for
running testcases, but the testsuite handles that in target-libpath.exp.

So, for stage2+, let's add just prev- libgcc.  We do this in both HOST_ and
TARGET_EXPORTS; when building stage2 host tools with the stage1 compiler, if
the stage1 compiler was built with the system gcc, it will load the stage1
libgcc instead of the system libgcc--but that also happens without this
change.  If that turns out to be a problem, we could arrange to only add it
to HOST_EXPORTS for stage3+ in a later patch.

With this change nothing is using TARGET_LIB_PATH, so let's remove it.

Many thanks to Iain Sandoe, who tested this on Darwin systems both with and
without system libstdc++, and reports it working.

	PR bootstrap/105688

ChangeLog:

	* Makefile.tpl (HOST_EXPORTS, BASE_TARGET_EXPORTS): Don't add
	TARGET_LIB_PATH to RPATH_ENVVAR.
	(TARGET_LIB_PATH): Remove.
	(HOST_LIB_PATH): Add HOST_LIB_PATH_gcc.
	(HOST_LIB_PATH_gcc): In prev-gcc, and only when bootstrapping.
	* Makefile.in: Regenerate.

iains committed:
aarch64: Register rng builtins with uint64_t pointers.

Currently, these are registered as unsigned_intDI_type_node which is not
necessarily the same type definition as uint64_t.  On platforms where these
differ that causes fails in consuming the arm_acle.h header.

gcc/ChangeLog:

	* config/aarch64/aarch64-builtins.cc (aarch64_init_rng_builtins):
	Register these builtins with a pointer to uint64_t rather than unsigned
	DI mode.

iains committed:
libstdc++, Darwin: Handle a linker warning [PR112397].

Darwin's linker warns when we make a direct branch to code that is
in a weak definition (citing that if a different implementation of
the week function is chosen by the dynamic linker this woule be an
error).

As the analysis in the PR shows, this can happen when we have hot/
cold partitioning and there is an error path that is primarily cold
but makes use of epilogue code in the hot section).  In this simple
case, we can easily deduce that the code is in fact safe; however
that is not something we can realistically implement in the linker.

Since the user-replaceable allocators are implemented using weak
definitions, this is a warning that is frequently flagged up in both
the testsuite and end-user code.

The chosen solution here is to suppress the hot/cold partitioning for
these cases (it is unlikely to impact performance much c.f. the
actual allocation).

	PR target/112397

libstdc++-v3/ChangeLog:

	* configure: Regenerate.
	* configure.ac: Detect if we are building for Darwin.
	* libsupc++/Makefile.am: If we are building for Darwin, then
	suppress hot/cold partitioning for the array allocators.
	* libsupc++/Makefile.in: Regenerated.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>
Co-authored-by: Jonathan Wakely <jwakely@redhat.com>

TamarChristinaArm committed:
AArch64: remove ls64 from being mandatory on armv8.7-a..

Hi All,

The Arm Architectural Reference Manual (Version J.a, section A2.9 on FEAT_LS64)
shows that ls64 is an optional extensions and should not be enabled by default
for Armv8.7-a.

This drops it from the mandatory bits for the architecture and brings GCC inline
with LLVM and the achitecture.

Note that we will not be changing binutils to preserve compatibility with older
released compilers.

Bootstrapped Regtested on aarch64-none-linux-gnu and no issues.

Ok for master? and backport to GCC 13,12,11?

Thanks,
Tamar

gcc/ChangeLog:

	* config/aarch64/aarch64-arches.def (AARCH64_ARCH): Remove LS64 from
	Armv8.7-a.

gcc/testsuite/ChangeLog:

	* g++.target/aarch64/acle/ls64.C: Add +ls64.
	* gcc.target/aarch64/acle/pr110100.c: Likewise.
	* gcc.target/aarch64/acle/pr110132.c: Likewise.
	* gcc.target/aarch64/options_set_28.c: Drop check for nols64.
	* gcc.target/aarch64/pragma_cpp_predefs_2.c: Correct header checks.

iains committed:
testsuite, aarch64, darwin: handle undefined symbols in a shared object

iains committed:
testsuite: hack to fix an lto test with insufficient function alignment.

iains committed:
libiberty: Fix error return value in pex_unix_exec_child [PR113957].

r14-5310-g879cf9ff45d940 introduced some new handling for spawning sub
processes.  The return value from the generic exec_child is examined
and needs to be < 0 to signal an error. However, the unix flavour of
this routine is returning the PID value set from the posix_spawn{p}.

This latter value is undefined per the manual pages for both Darwin
and Linux, and it seems Darwin, at least, sets the value to some
usually positive number (presumably the PID that would have been used
if the fork had succeeded).

The fix proposed here is to set the pid = -1 in the relevant error
paths.

	PR other/113957

libiberty/ChangeLog:

	* pex-unix.c (pex_unix_exec_child): Set pid = -1 in the error
	paths, since that is used to signal an erroneous outcome for
	the routine.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
testsuite: Temporary test skips for PR113150

gcc/testsuite/ChangeLog:

	* c-c++-common/analyzer/fd-glibc-byte-stream-socket.c:
	* c-c++-common/analyzer/fd-manpage-getaddrinfo-client.c:
	* c-c++-common/analyzer/fd-mappage-getaddrinfo-server.c:
	* c-c++-common/analyzer/fd-symbolic-socket.c:
	* gcc.dg/analyzer/fd-glibc-byte-stream-connection-server.c:

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
testsuite, g++: Temporary XFAILs for PR111081.

gcc/testsuite/ChangeLog:

	* g++.dg/opt/icf1.C:
	* g++.dg/opt/icf2.C:
	* g++.dg/opt/icf3.C:

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
testsuite,g++: Temporary XFAILs for PR112294/106435

gcc/testsuite/ChangeLog:

	* g++.dg/tls/thread_local13.C:
	* g++.dg/tls/thread_local14.C:

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
aarch64: Allow aarch64-linux-muscl for heap trampolines [PR113971].

This allows the same trampoline pattern to be used on all linux variants
rather than restricting it to linux gnu.

	PR target/113971.

libgcc/ChangeLog:

	* config/aarch64/heap-trampoline.c: Allow all linux variants.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
libgcc, aarch64: Allow for BE platforms in heap trampolines.

This arranges that the byte order of the instruction sequences is
independent of the byte order of memory.

libgcc/ChangeLog:

	* config/aarch64/heap-trampoline.c
	(aarch64_trampoline_insns): Arrange to encode instructions as a
	byte array so that the order is independent of memory byte order.
	(struct aarch64_trampoline): Likewise.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

iains committed:
posted

iains committed:
shortq

iains commented:
do you think this is something we should try to get into our gcc-14 release?
(if so, it presumably needs rebasing)

fxcoudert commented:
OK. We should probably add the M3 there. But it's a rat's race, and we should have a sensible default. What happens if someone passes `native` and we don't detect the sysctl value?

iains commented:
> OK. We should probably add the M3 there. 

I'll add M3 to the known CPUs shortly (done locally, but I need to rebase and push)

> But it's a rat's race, and we should have a sensible default. What happens if someone passes `native` and we don't detect the sysctl value?

that's a general question about error recovery/fallback - what does the linux impl. do if it encounters an unknown cpu in /proc/.... ?

we could have a fallback to m1, I suppose ..  getting early access to h/w info seems unlikely given that it's likely to be held close prior to any release.

fxcoudert committed:
Add M3, remove unused code

fxcoudert commented:
- Rebased, 
- Added M3 variants (all named M3 because I don't think we want to have more detail)
- Removed older, unused variants

Can't test right now, sorry

iains commented:
response from IRC:  hmm it should only return null on error, when we simply didn't find something we know it returns DEFAULT_ARCH + extensions https://sourcegraph.com/github.com/gcc-mirror/gcc@7d8585c0c0e5934780281abdee256ae6553e56e8/-/blob/gcc/config/aarch64/driver-aarch64.cc?L402

I will take a look at this when adding the patch to the branch (I know you are also busy).

fxcoudert committed:
If CPU type is unknown, return M1

iains commented on gcc/config/aarch64/driver-aarch64.cc:
```diff
@@ -28,6 +28,95 @@
+  if (!arch && !tune && !cpu)
+    return NULL;
+
+  sysctlbyname("hw.cpufamily", &family, &len, NULL, 0);
```
I think we should check that the sysctlbyname succeeds and issue a diagnostic if not.  For the sake of "doing the same thing as Linux" returning NULL would also be reasonable there.

iains reviewed (commented):
I've squashed the three patches here onto my current rebase (which should be ready tomorrow - without making any other changes).

iains commented:
https://github.com/iains/gcc-darwin-arm64/commit/9bc66f9b11baf1c291b2afac429cb12bf8461f91 includes the content of this branch.

iains commented:
the merged patch has been tested on M1 and M3 - so I think we're all good with this - and maybe a couple of minor tweaks to the code for error checking.

fxcoudert commented:
The other uses of this `sysctlbyname` in the darwin code of GCC expect that it never fails, so I think we're safe, but I'll open a PR to add error checking at some point.

iains commented:
I have merged the patch here and added one to select `apple-m1` as a default CPU if there is no configure one.
